### PR TITLE
egnine/sync-manager: add config support

### PR DESCRIPTION
**Pull request by gloriousCode**
# PR Description
This is a quick & basic QoL improvement which adds the sync manager config to our config file. No longer do you need to add so many params when running the application to change behaviour. You can still use the running params to override your config.

Also, the sync manager is a pretty trustworthy subsystem now, so I've added the ability to modify what logging events to output. You can toggle whether you want to see:
- Switch protocol updates
- Initial sync updates
- Event updates

Errors, default subsystem logs all still show and Verbose is still a thing. 

Please delete options that are not relevant and add an `x` in `[]` as item is complete.
- [x] New feature (non-breaking change which adds functionality)

## How has this been tested
- [x] go test ./... -race
- [x] golangci-lint run
- [x] `TestGetDefaultSyncManagerConfig()`

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -696,6 +696,56 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+	}
+	if c.SyncManagerConfig.TimeoutWebsocket <= 0 {
+		log.Warnf(log.ConfigMgr, "invalid sync manager websocket timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutWebsocket, DefaultSyncerTimeoutWebsocket)
+		c.SyncManagerConfig.TimeoutWebsocket = DefaultSyncerTimeoutWebsocket
+	}
+	if c.SyncManagerConfig.PairFormatDisplay == nil {
+		log.Warnf(log.ConfigMgr, "invalid sync manager pair format value %v, defaulting to %v\n", c.SyncManagerConfig.PairFormatDisplay, c.CurrencyPairFormat)
+		c.SyncManagerConfig.PairFormatDisplay = c.CurrencyPairFormat
+	}
+	if c.SyncManagerConfig.TimeoutREST <= 0 {
```
Comment (shazbert) [inline config/config.go:739]: if I purge fiat display currency it purges and never comes back and stops the sync manager from working🙏

File: engine/helpers.go [hunk 2]
```diff
@@ -185,19 +185,27 @@ func (bot *Engine) SetSubsystem(subSystemName string, enable bool) error {
...
-					SynchronizeOrderbook:    bot.Settings.EnableOrderbookSyncing,
-					SynchronizeTrades:       bot.Settings.EnableTradeSyncing,
-					SynchronizeContinuously: bot.Settings.SyncContinuously,
-					TimeoutREST:             bot.Settings.SyncTimeoutREST,
-					TimeoutWebsocket:        bot.Settings.SyncTimeoutWebsocket,
-					NumWorkers:              bot.Settings.SyncWorkersCount,
-					FiatDisplayCurrency:     bot.Config.Currency.FiatDisplayCurrency,
-					Verbose:                 bot.Settings.Verbose,
+				cfg := bot.Config.SyncManagerConfig
+				cfg.SynchronizeTicker = bot.Settings.EnableTickerSyncing && cfg.SynchronizeTicker
```
Comment (shazbert) [inline engine/helpers.go:189]: When ticker sync is enabled in config this pattern used to override if set using `flagSet.WithBool` functionality in engine.
```
.\gocryptotrader -orderbooksync=true -tickersync=false
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Looks good. Minor nits.

File: engine/helpers.go (hunk 2 above)
Comment (gloriousCode) [reply engine/helpers.go:189]: Could you rephrase that please?

File: config/config.go (hunk 1 above)
Comment (gloriousCode) [reply config/config.go:739]: Well don't do that 😆 Will look into it

File: engine/helpers.go (hunk 2 above)
Comment (shazbert) [reply engine/helpers.go:189]: When ticker sync is set to true in config, when I try to set it to false via command line it will not. I don't english well.

File: config/config.go (hunk 1 above)
Comment (shazbert) [reply config/config.go:742]: This is still an issue, when I remove this `fiatdisplayCurrency` from the config it stops the starting of the `syncMnager`

File: config/config.go [hunk 3]
```diff
@@ -696,6 +696,59 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+	if c.SyncManagerConfig == (SyncManagerConfig{}) {
+		c.SyncManagerConfig = GetDefaultSyncManagerConfig()
+		return
+	}
+	if c.SyncManagerConfig.TimeoutWebsocket <= 0 {
+		log.Warnf(log.ConfigMgr, "invalid sync manager websocket timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutWebsocket, DefaultSyncerTimeoutWebsocket)
+		c.SyncManagerConfig.TimeoutWebsocket = DefaultSyncerTimeoutWebsocket
+	}
+	if c.SyncManagerConfig.PairFormatDisplay == nil {
+		log.Warnf(log.ConfigMgr, "invalid sync manager pair format value %v, using default format eg BTC-USD\n", c.SyncManagerConfig.PairFormatDisplay)
```
Comment (shazbert) [inline config/config.go:736]: 👍

File: engine/helpers.go (hunk 2 above)
Comment (shazbert) [reply engine/helpers.go:189]: For the inverse if I have both set to false in config and I run this command, it doesn't work:

```
 .\gocryptotrader --orderbooksync=true --tickersync=true
```

**Review by shazbert — CHANGES_REQUESTED** [review]
Thanks for changes, just 2 things I noticed.

File: engine/helpers.go (hunk 2 above)
Comment (gloriousCode) [reply engine/helpers.go]: Default app param values have now become `false` to address this

Comment (gloriousCode) [reply engine/helpers.go]: Resolved, I forgot about the lovely existing helper functions

**Review by shazbert — APPROVED** [review]
LGTM Thanks!

File: config/config.go [hunk 4]
```diff
@@ -696,6 +696,63 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+// sets defaults if values are invalid
+func (c *Config) CheckSyncManagerConfig() {
+	m.Lock()
+	defer m.Unlock()
+	if c.SyncManagerConfig == (SyncManagerConfig{}) {
+		c.SyncManagerConfig = GetDefaultSyncManagerConfig()
+		return
+	}
+	if c.SyncManagerConfig.TimeoutWebsocket <= 0 {
+		log.Warnf(log.ConfigMgr, "invalid sync manager websocket timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutWebsocket, DefaultSyncerTimeoutWebsocket)
```
Comment (thrasher-) [inline config/config.go:732]:
```suggestion
		log.Warnf(log.ConfigMgr, "Invalid sync manager websocket timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutWebsocket, DefaultSyncerTimeoutWebsocket)
```

File: config/config.go [hunk 5]
```diff
@@ -696,6 +696,63 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+		c.SyncManagerConfig.TimeoutWebsocket = DefaultSyncerTimeoutWebsocket
+	}
+	if c.SyncManagerConfig.PairFormatDisplay == nil {
+		log.Warnf(log.ConfigMgr, "invalid sync manager pair format value %v, using default format eg BTC-USD\n", c.SyncManagerConfig.PairFormatDisplay)
```
Comment (thrasher-) [inline config/config.go:736]:
```suggestion
		log.Warnf(log.ConfigMgr, "Invalid sync manager pair format value %v, using default format eg BTC-USD\n", c.SyncManagerConfig.PairFormatDisplay)
```

File: config/config.go [hunk 6]
```diff
@@ -696,6 +696,63 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+		c.SyncManagerConfig.PairFormatDisplay = &currency.PairFormat{
+			Uppercase: true,
+			Delimiter: currency.DashDelimiter,
+		}
+	}
+	if c.SyncManagerConfig.TimeoutREST <= 0 {
+		log.Warnf(log.ConfigMgr, "invalid sync manager REST timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutREST, DefaultSyncerTimeoutREST)
```
Comment (thrasher-) [inline config/config.go:743]:
```suggestion
		log.Warnf(log.ConfigMgr, "Invalid sync manager REST timeout value %v, defaulting to %v\n", c.SyncManagerConfig.TimeoutREST, DefaultSyncerTimeoutREST)
```

File: config/config.go [hunk 7]
```diff
@@ -696,6 +696,63 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+		c.SyncManagerConfig.TimeoutREST = DefaultSyncerTimeoutREST
+	}
+	if c.SyncManagerConfig.NumWorkers <= 0 {
+		log.Warnf(log.ConfigMgr, "invalid sync manager worker count value %v, defaulting to %v\n", c.SyncManagerConfig.NumWorkers, DefaultSyncerWorkers)
```
Comment (thrasher-) [inline config/config.go:747]:
```suggestion
		log.Warnf(log.ConfigMgr, "Invalid sync manager worker count value %v, defaulting to %v\n", c.SyncManagerConfig.NumWorkers, DefaultSyncerWorkers)
```

File: config/config.go [hunk 8]
```diff
@@ -696,6 +696,63 @@ func (c *Config) GetAvailablePairs(exchName string, assetType asset.Item) (curre
...
+		c.SyncManagerConfig.NumWorkers = DefaultSyncerWorkers
+	}
+	if c.SyncManagerConfig.FiatDisplayCurrency.IsEmpty() {
+		log.Warnf(log.ConfigMgr, "invalid sync manager fiat display currency value, defaulting to %v\n", currency.USD)
```
Comment (thrasher-) [inline config/config.go:751]:
```suggestion
		log.Warnf(log.ConfigMgr, "Invalid sync manager fiat display currency value, defaulting to %v\n", currency.USD)
```

File: engine/sync_manager.go [hunk 9]
```diff
@@ -196,19 +195,25 @@ func (m *syncManager) Start() error {
...
+				createdCounter)
+		}
 		m.initSyncStartTime = time.Now()
 	}
 
 	go func() {
 		m.initSyncWG.Wait()
 		if atomic.CompareAndSwapInt32(&m.initSyncCompleted, 0, 1) {
-			log.Debugf(log.SyncMgr, "Exchange CurrencyPairSyncer initial sync is complete.")
+			if m.config.LogInitialSyncEvents {
```
Comment (thrasher-) [inline engine/sync_manager.go:209]: These two if statements can be squished into one

**Review by thrasher- — CHANGES_REQUESTED** [review]
Nice! Just some minor nitters

**Review by thrasher- — APPROVED** [review]
Like a glorious beef stroganoff!

**Review by shazbert — APPROVED** [review]
Ch ACK
